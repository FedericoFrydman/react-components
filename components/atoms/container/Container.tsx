import { forwardRef, Ref, ReactElement } from 'react';
import cn from 'classnames';

import './container.scss';

export type ContainerBgColors = 'light-gray' | 'white' | 'none';

export interface ContainerProps extends Omit<React.HTMLProps<HTMLDivElement>, 'css'> {
  /**
   * Define if container has background
   * @default 'white'
   */
  background?: ContainerBgColors;
  /**
   * Define if container has position fixed
   */
  fixed?: boolean;
  /**
   * Define if container has display flex
   */
  flexbox?: boolean;
  /**
   * Define the flex-direction of flex-container
   */
  flexDirection?: 'row' | 'row-reverse' | 'column' | 'column-reverse';
  /**
   * Define the flex-wrap of flex-container
   */
  flexWrap?: 'wrap' | 'nowrap' | 'wrap-reverse';
  /**
   * Define the flex-direction
   */
  justifyContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around';
  /**
   * Define the align-items for flex-container
   */
  alignItems?: 'flex-start' | 'flex-end' | 'center' | 'baseline' | 'stretch';
  /**
   * Define the flex-direction
   */
  alignContent?: 'flex-start' | 'flex-end' | 'center' | 'space-between' | 'space-around' | 'stretch';
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Children Container
   */
  children?: string | number | ReactElement | ReactElement[];
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLDivElement>;
}

const ContainerComponent = (
  {
    background = 'white',
    classNames = [],
    children,
    fixed,
    flexbox,
    flexDirection,
    flexWrap,
    justifyContent,
    alignItems,
    alignContent,
    ...props
  }: ContainerProps,
  ref: ContainerProps['ref'],
) => {
  const classContainer = 'lsux-container';
  const cnContainer = cn(
    {
      [`${classContainer}`]: true,
      [`${classContainer}--${background}`]: background !== 'none',
      [`${classContainer}--fixed`]: fixed,
      [`${classContainer}--flexbox`]: flexbox,
      [`${classContainer}--flex-direction-${flexDirection}`]: flexDirection,
      [`${classContainer}--flex-wrap-${flexWrap}`]: flexWrap,
      [`${classContainer}--flex-justify-${justifyContent}`]: justifyContent,
      [`${classContainer}--flex-items-${alignItems}`]: alignItems,
      [`${classContainer}--flex-content-${alignContent}`]: alignContent,
    },
    classNames,
  );

  return (
    <div {...props} className={cnContainer} ref={ref}>
      {children}
    </div>
  );
};

export const Container = forwardRef<HTMLDivElement, ContainerProps>(ContainerComponent);
