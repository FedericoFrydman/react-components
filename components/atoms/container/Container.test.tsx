import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { Container, ContainerProps } from './Container';

describe('Container Component', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Container />);
  });

  test('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('Container Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Container />);
    testId = 'test-container';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderContainer({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLDivElement>();

    it('provides a ref to the spinner', () => {
      expect(renderContainer({ ref })).toEqual(ref.current);
    });
  });

  const renderContainer = (props: ContainerProps = {}): HTMLElement => {
    render(<Container {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
