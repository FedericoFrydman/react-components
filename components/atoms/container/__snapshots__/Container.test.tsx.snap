// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Component should render the component 1`] = `
<div
  className="lsux-container lsux-container--white"
/>
`;

exports[`Container Component should render the component 2`] = `
<div
  className="lsux-container lsux-container--white"
/>
`;
